hubcast/github/handler.py

```diff
@@ -0,0 +1,142 @@
+import asyncio
+import os
+import sys
+import traceback
+from subprocess import CalledProcessError
+from typing import Dict
+
+import aiohttp
+from aiohttp import web
+from gidgethub import aiohttp as gh_aiohttp
+from gidgethub import sansio
+from gidgetlab import aiohttp as gl_aiohttp
+
+from hubcast.account_map.abc import AccountMap
+from hubcast.github import config
+from hubcast.github.auth import GitHubAuthenticator
+from hubcast.github.routes import router
+from hubcast.gitlab.auth import GitLabAuthenticator
+from hubcast.utils.git import Git
+
+
+async def create_repo(
+    git: Git,
+    repo_lock: asyncio.Lock,
+    repo_path: str,
+    github_url: str,
+    gitlab_url: str,
+):
+    await repo_lock.acquire()
+    try:
+        if not os.path.exists(repo_path):
+            os.makedirs(repo_path)
+            git("init")
+            git(f"remote add github {github_url}")
+            git(f"remote add gitlab {gitlab_url}")
+
+    except CalledProcessError:
+        os.rmdir(repo_path)
+
+    finally:
+        repo_lock.release()
+
+
+class GitHubHandler:
+    def __init__(
+        self,
+        gh_authenticator: GitHubAuthenticator,
+        gh_requester: str,
+        webhook_secret: str,
+        repos_path: str,
+        account_map: AccountMap,
+        gl_authenticator: GitLabAuthenticator,
+        gl_instance_url: str,
+    ):
+        self.auth = gh_authenticator
+        self.requester = gh_requester
+        self.webhook_secret = webhook_secret
+        self.repos_path = repos_path
+        self.account_map = account_map
+        self.gl_auth = gl_authenticator
+        self.gl_instance_url = gl_instance_url
+        self.repo_locks: Dict[str, asyncio.Lock] = {}
+
+    async def handle(self, request):
+        try:
+            # read the GitHub webhook payload
+            body = await request.read()
+            event = sansio.Event.from_http(
+                request.headers, body, secret=self.webhook_secret
+            )
+            print("GH delivery ID", event.delivery_id, file=sys.stderr)
+
+            # get the owner and name of the repo which sent the event
+            repo_data = event.data["repository"]
+            [repo_owner, repo_name] = repo_data["full_name"].split("/", 1)
+            repo_path = f"{self.repos_path}/{repo_owner}/{repo_name}"
+
+            # create a repo lock if it doesn't already exist
+            if repo_path not in self.repo_locks:
+                self.repo_locks[repo_path] = asyncio.Lock()
+
+            repo_lock = self.repo_locks[repo_path]
+
+            # put account map conversation here
+            try:
+                github_user = event.data["sender"]["login"]
+                gitlab_user = self.account_map(github_user)
+            except Exception as exc:
+                print(exc)
+                return web.Response(status=200)
+
+            if gitlab_user is None:
+                print("Unauthorized User")
+                return web.Response(status=200)
```

Comment:
Should this be a 403?

Reply:
No this'll be true for most external PRs that aren't created by folks with LC accounts. We don't want to return a 4xx for these because that'd likely cause the app webhook to be disabled or for the event to be sent again. (See the [GitHub Docs](https://docs.github.com/en/webhooks/using-webhooks/best-practices-for-using-webhooks#respond-within-10-seconds) & [GitLab Docs](https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#configure-your-webhook-receiver-endpoint).) From GitHub's perspective these should be silent failures since it's not an error with the configuration of the program but rather an acceptable end state.